spmd/tensor/parallel/api.py

```diff
@@ -1,122 +1,195 @@
-            output.redistribute(output.device_mesh, replicate)
-            .to_local()
-            .contiguous()
-        )
+def parallelize_module(
+    module: nn.Module,
+    device_mesh: DeviceMesh,
+    parallelize_plan: Union[ParallelStyle, Dict[str, ParallelStyle]],
+    tp_mesh_dim: int = 0,
+) -> nn.Module:
+    """
+    The major API for tensor parallelism in PyTorch. We parallelize module
+    or sub_modules based on a parallelize_plan which contains the parallel_style
+    which indicates how user want the module or sub_module to be parallelized.
+    User can also specify different parallel_style per FQN. The API supports 2D
+    parallelism natively by accepting an n-dimension device_mesh and users just
+    need to specify the dimension where we perform tensor parallelism on.
+
+    Args:
+        module (nn.Module):
+            :class:``nn.Module`` object to be parallelized.
+        device_mesh (DeviceMesh):
+            :class:``DeviceMesh`` object which describes the mesh topology
+            of devices for the DTensor.
+        parallelize_plan (Union[ParallelStyle, Dict[str, ParallelStyle]]):
+            The plan used to parallelize the module. It can be either a
+            :class:``ParallelStyle`` object which contains how
+            we prepare input/output for Tensor Parallelism or it can be a
+            dict of FQN and its corresponding :class:``ParallelStyle`` object.
+        tp_mesh_dim (int):
+            the dimension of ``device_mesh`` where we perform
+            Tensor Parallelism on.
 
+    Return:
+        A :class:``nn.Module`` object parallelized.
 
-def tp_shard_self_attn(
-    name: str, module: nn.Module, device_mesh: DeviceMesh
-) -> None:
-    col_wise_sharding: Sequence[Placement] = [Shard(0)]
-    row_wise_sharding: Sequence[Placement] = [Shard(1)]
-    replicate: Sequence[Placement] = [Replicate()] * device_mesh.ndim
-
-    def _shard_self_attn_params(name: str, module: nn.Module) -> None:
-        if isinstance(module, nn.Linear):
-            if name == "qkv":
-                sharded_weight = nn.Parameter(
-                    distribute_tensor(
-                        module.weight, device_mesh, col_wise_sharding
-                    )
+    Example::
+        >>> # xdoctest: +SKIP("distributed")
+        >>> from from spmd.tensor.parallel import parallelize_module, PairwiseParallel
+        >>>
+        >>> # Define the module.
+        >>> m = Model(...)
+        >>> m = parallelize_module(m, PairwiseParallel())
+        >>>
+
+    .. warning::
+        ``PairwiseParallel`` comes with constraints for now. If you need finer
+        granularity 
+    """
+
+    if device_mesh.ndim > 1:
+        device_mesh = _create_1d_device_mesh(device_mesh, tp_mesh_dim)
+
+    if isinstance(parallelize_plan, ParallelStyle):
+        if _is_mha_for_pairwise_parallel(module):
+            return _parallelize_multihead_attn(module, device_mesh)
+        elif _is_mlp_for_pairwise_parallel(module):
```

Comment:
does this mean if `parallelize_plan` is ParallelStyle, we only support PairwiseParallel? what if user pass in a single RowwiseParallel or ColwiseParallel? shall we just error out in those case?

Reply:
Oh that is TBD since Xilun's PR has not been merged yet. Let me add a TODO here to clarify.